### ISSUE TITLE
Recluster the surface audit with aprender-core: a real sklearn-parity falsifier, and contracts-pv's first coverage

### ISSUE THREAD

**Issue by noahgift** [paiml/aprender#2591]
The clustering that now determines where aprender's gates go was produced with
**scikit-learn** — using capabilities aprender already ships. That is the finding.

## Verified at HEAD (4bbfeb07f)

| capability | location | verdict |
|---|---|---|
| `KMeans` (+ `.inertia()`) | `crates/aprender-core/src/cluster/kmeans.rs:50` — `pub struct KMeans {` | ✅ |
| `silhouette_score` | `crates/aprender-core/src/metrics/mod.rs:328` | ✅ |
| `TfIdfVectorizer` | `crates/aprender-core/src/text/vectorize/` — exported from `mod.rs` (**not** `tfidf_vectorizer.rs`; that path does not exist) | ✅ path corrected |
| `HashingVectorizer` | `crates/aprender-core/src/text/vectorize/hashing_vectorizer.rs` | ✅ |
| k-sweep by silhouette | `crates/aprender-core/examples/automl_clustering.rs` (**not** `examples/` at repo root) | ✅ path corrected |
| knee/elbow detection | **`elbow` appears in 0 `.rs` files across the workspace** | ✅ **the one real gap** |

Everything except knee detection is a **wiring job**, not new capability.

## Why this is load-bearing, not vanity dogfooding

P1 is sklearn parity, and today it is asserted against synthetic fixtures and `iris`.
Reclustering the 830-row surface audit with `aprender-core` and diffing against the
sklearn run is **a real sklearn-parity falsifier on a real sparse, high-dimensional,
badly-conditioned matrix** — nothing like iris.

- If `apr`'s KMeans does not reproduce it, that is a **P1 defect the framework's own
  release process found.**
- If it does, parity becomes **evidence rather than assertion.**

aprender's false negatives are fleet-multiplied, so aprender's gates must pass its own
doctrine.

## The circularity, and why the fix is differential

If aprender's clustering determines gate placement, a KMeans bug corrupts gate allocation
**silently** — the same coupled-failure shape as benchmark-target-equals-release-artifact
in `APR-BENCH-RFC-001`.

So this must be **differential, not replacement**. Both implementations run; the gate
asserts agreement:

- **ARI**(sklearn labels, apr labels) ≥ threshold
- **inertia** within tolerance at fixed `k`, fixed seed
- **silhouette** within tolerance

sklearn stays as the **differential oracle** — the role ollama plays in `F-DIFF-*`. It is
not in the render path, so this does not violate sovereignty; it is the external check
that makes the sovereign implementation **falsifiable**.

## One ticket, two zeros closed

`contracts-pv` is **76 features at 0% coverage** (#2589) — the tool that enforces provable
contracts across the stack has no contract on itself. Expressing the clustering invariants
as a `pv` contract gates the clustering **and** becomes `contracts-pv`'s first dogfood
coverage.

`contracts/apr-dogfood-cluster-v1.yaml`:

| invariant | falsifier |
|---|---|
| ARI(sklearn, apr) ≥ threshold at fixed k, fixed seed | perturb one centroid init → RED |
| inertia monotone non-increasing in k | inject a k where it rises → RED |
| silhouette on cosine matches reference | swap the metric → RED |
| every feature assigned exactly one cluster | drop an assignment → RED |
| cluster count ≥ floor (vacuity) | force k=1 → RED |
| label→id map stable across re-runs on identical input | permute ids → RED |

That last one mechanizes the trap: **k-means ids permute**, so the contract pins
`cluster_label` as the durable key and `cluster_id` as incidental.

## Work

1. **`apr cluster surface`** — new subcommand. TF-IDF + KMeans + silhouette from
   `aprender-core`, emits `surface_audit_clustered.csv`. Lands in `apr-core-commands`.
2. **Knee detection in `aprender-core`** — the real gap. `automl_clustering.rs` picks `k`
   by max silhouette; on this surface silhouette **climbs to k=44 without settling** and
   gives no answer at all. Max-distance-to-chord is ~20 lines and is a genuine sklearn
   **non-parity win** — sklearn has no knee finder either. **Beat, not parity.**
   Note also: `automl_clustering.rs` caps at **k ≤ 10**, so it would have missed the knee
   at **k=14** entirely.
3. **`pv` contract** — the six invariants above.
4. **Differential gate** — sklearn oracle, ARI assertion, wired into Tier 2 beside
   `F-META-*`.

## Blocking prerequisite — three KMeans implementations

```
crates/aprender-core/src/cluster/kmeans.rs          pub struct KMeans
crates/aprender-train/src/aprender_compat.rs:191    pub struct KMeans
crates/aprender-verify-ml/src/ml/active_learner.rs  KMeansClustering (+ its own silhouette_score)
```

Three implementations is a divergence hazard of **exactly the G0.5 duplicate-binary
shape** (#2585), where two hand-duplicated `apr` bins had already diverged on the `dhat`
allocator. **Determine which is canonical before any of them becomes load-bearing** — if
gate placement depends on KMeans, it must depend on a named one.

Refs #2582, #2589, #2585

**Comment by noahgift** [comment]
## APR-CLUSTER-001 filed — 8 tickets, milestone, labels

Feature-set spec landed as issues in dependency order, milestone **APR-CLUSTER-001** (#2).

| ticket | # | labels |
|---|---|---|
| CLUS-000 canonicalize `KMeans` | **#2592** | `blocking` |
| CLUS-001 knee detection | #2593 | `beat` |
| CLUS-002 k-sweep, cap removed | #2594 | |
| CLUS-003 `apr cluster surface` | #2595 | |
| CLUS-004 `pv` contract | **#2596** | `dogfood-coverage` `needs-mutation` |
| CLUS-005 sklearn differential gate | #2597 | `needs-mutation` |
| CLUS-006 per-cluster coverage floor | #2598 | `dogfood-coverage` `needs-mutation` |
| CLUS-007 sibling-sweep protocol | #2599 | |

**#2592 blocks everything downstream** and is correctly labelled so. Three `KMeans`
implementations — `aprender-core/src/cluster/kmeans.rs`,
`aprender-train/src/aprender_compat.rs:191`, `aprender-verify-ml/src/ml/active_learner.rs`
(`KMeansClustering`, with its own `silhouette_score`) — all verified present at HEAD. If
gate placement is going to depend on KMeans, it must depend on a **named** one.

The spec's own do-not-build entry is the sharp version of why: **"consolidate KMeans
without first proving the three agree"** hides a pre-existing divergence. That is exactly
what #2585 found in the duplicated `apr` binaries, where the two had already parted ways
on the `dhat` allocator and nobody knew.

### The do-not-build list is the most valuable part of this spec

Nine entries, each naming a plausible-looking move that would make things worse:

- **Replace sklearn instead of running both** — couples the failure modes; a KMeans bug
  would silently corrupt gate allocation
- **Bake `cluster_id` into a contract or gate key** — ids permute; the contract silently
  re-points
- **Report cluster coverage without feature coverage** — vacuity one level up
- **Derive `quality_1_10` from cluster membership** — clustering is a prior, not evidence
- **"Keep the three KMeans in sync"** — that is the state they are already failing at
- **Put the sklearn oracle in the unit-test path** — regresses the working hermetic setup
- **Select k by max silhouette alone** — returns no answer on this input (0.26 at k=44,
  still climbing)
- **Any hardcoded k ceiling** — the current one already hides the correct answer

That last pair is measured, not theoretical: `automl_clustering.rs` caps at **k ≤ 10** and
would have missed the knee at **k=14** outright, while silhouette never settles. The
existing tooling gives *no answer* on this input — which is why CLUS-001 is labelled
`beat` rather than parity. **sklearn has no knee finder either.**

Capability claims re-verified at HEAD before filing; two paths in the source spec were
corrected (`TfIdfVectorizer` is exported from `text/vectorize/mod.rs`, not a
`tfidf_vectorizer.rs`; `automl_clustering.rs` lives under `crates/aprender-core/examples/`,
not repo-root `examples/`). `elbow` confirmed absent from all `.rs` in the workspace.

**Comment by noahgift** [comment]
triage 2026-09-23 (aprender-dd): milestoned to 0.76.0, theme Classical ML + AutoML. It is the classical-ML parity work (#3146 epic, CLUS rows), which no 0.70–0.75 theme covered. Operator selection 2026-09-23 ("New 0.76 release theme", wording by the cop aprender-cf); moved out of `backlog`.

**Comment by noahgift** [comment]
Closed by paiml-implement — folded into epic #4164 (triage evidence/triage-1717.md)
